openlr/candidate_points_getter.cpp

```diff
@@ -0,0 +1,97 @@
+#include "openlr/candidate_points_getter.hpp"
+#include "openlr/helpers.hpp"
+#include "openlr/timer.hpp"
+
+#include "routing/road_graph.hpp"
+
+#include "storage/country_info_getter.hpp"
+
+using namespace routing;
+
+namespace
+{
+MwmSet::MwmId GetMwmIdByPoint(m2::PointD const & p, Index const & index)
+{
+  // TODO(mgsergio): Should I cache it later?
+  auto const infoGetter =
+      storage::CountryInfoReader::CreateCountryInfoReaderTwoComponentMwms(GetPlatform());
+  auto const & countryId = infoGetter->GetRegionCountryId(p);
+  ASSERT(!countryId.empty(), ("CountryId is \"\" for point", MercatorBounds::ToLatLon(p)));
+  auto const & countryFile = platform::CountryFile(countryId);
+  return index.GetMwmIdByCountryFile(countryFile);
+}
+}  // namespace
+
+namespace openlr
+{
+void CandidatePointsGetter::GetJunctionPointCandidates(m2::PointD const & p,
+                                                       vector<m2::PointD> & candidates)
+{
+  Timer t(m_stat.m_pointCadtidatesTime);
+
+  size_t const kRectSideMeters = 110;
+
+  auto const mwmId = GetMwmIdByPoint(p, m_index);
+
+  auto const rect = MercatorBounds::RectByCenterXYAndSizeInMeters(p, kRectSideMeters);
+  auto const selectCandidates = [&rect, &candidates](FeatureType & ft) {
+    ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+    ft.ForEachPoint(
+        [&rect, &candidates](m2::PointD const & candidate) {
+          if (rect.IsPointInside(candidate))
+            candidates.emplace_back(candidate);
+        },
+        scales::GetUpperScale());
+  };
+
+  m_index.ForEachInRect(selectCandidates, rect, scales::GetUpperScale());
+
+  // TODO: Move this to a separate stage.
+  // 1030292476 Does not match. Some problem occur with poins.
+  // Either points duplicatate or something alike. Check thils
+  // later. The idea to fix this was to move SortUnique to the stage
+  // aftex enriching with projections.
+
+  my::SortUnique(candidates,
+                 [&p](m2::PointD const & a, m2::PointD const & b) {
+                   return MercatorBounds::DistanceOnEarth(a, p) <
```

Comment:
Beware there! You calculate distances dynamically, and due to bugs in compilers these computations may be instable. I.e. if sometimes wide floating point registers are used (wider than 8 bytes), over-optimizing compiler may deduce during sorting that
```
a = b (when 8-bytes registers are used)
b < c (when 16-bytes registers are used)
c < a (when 8-bytes registers are used)
```

Reply:
For example - https://gcc.gnu.org/bugzilla/show_bug.cgi?id=323